janus.go

```diff
@@ -313,79 +310,94 @@ type Session struct {
 	gateway *Gateway
 }
 
-func (session *Session) send(msg map[string]interface{}, transaction chan interface{}) {
-	msg["session_id"] = session.ID
-	session.gateway.send(msg, transaction)
+func (session *Session) send(msg any, transactionID string, transaction chan interface{}) {
+	session.gateway.send(msg, transactionID, transaction)
 }
 
 // Attach sends an attach request to the Gateway within this session.
 // plugin should be the unique string of the plugin to attach to.
 // On success, a new Handle will be returned and error will be nil.
-func (session *Session) Attach(plugin string) (*Handle, error) {
-	req, ch := newRequest("attach")
-	req["plugin"] = plugin
-	session.send(req, ch)
-
-	var success *SuccessMsg
-	msg := <-ch
-	switch msg := msg.(type) {
-	case *SuccessMsg:
-		success = msg
-	case *ErrorMsg:
-		return nil, msg
-	}
+func (session *Session) AttachSession(ctx context.Context, msg BaseMsg) (*SuccessMsg, error) {
+	ch := make(chan any)
+	session.send(msg, msg.ID, ch)
+
+	select {
+	case response := <-ch:
+		switch res := response.(type) {
+		case *SuccessMsg:
+			handle := new(Handle)
+			handle.session = session
+			handle.ID = res.Data.ID
+			handle.Events = make(chan interface{}, 8)
+			session.Lock()
+			session.Handles[handle.ID] = handle
+			session.Unlock()
+			return res, nil
+		case *ErrorMsg:
+
+			return nil, res
+		default:
+			return nil, nil
 
-	handle := new(Handle)
-	handle.session = session
-	handle.ID = success.Data.ID
-	handle.Events = make(chan interface{}, 8)
+		}
```

Comment:
Same here, the message will most probably will arrive later and will mess up the communication.